Add Facets foundation (facet id/label/score) with JSON serialization

uklotzde opened the pull request:
As promised we could start introducing the foundation for *Facets* (e.g. *Track Facets*, formerly known as *Custom Tags*) without touching the database schema or UI aspects yet.

Each *facet* is represented by a *triple*:  a *facet id*, a *label*, and a *score*:

- The *facet id* is a lowercase, non-empty ASCII string without whitespace
- The *label* is non-empty free text
- The score is a floating-point value in the interval [0.0, 1.0]
 
Both *facet id* and *label* are optional (= nullable), but at least one of them must be present. The *score* is mandatory and is supposed to be interpreted as a *weight*, *degree of cohesion*, or *normalized level* with a default value of 1.0.

The following combinations are allowed, missing values are indicated by `-`:
- Plain Tag: `(-, label, score)`, e.g. `label` = "Top 40", `score` = 1.0 for tagging a track with the label "Top 40" and full (= no particular, i.e. binary) score. Instead of using a score of 0.0 you would simply remove the entire tag then.
- Faceted Score: `(facet id, -, score)`, e.g. `facet id` = "energy", `score` = 0.8 for tagging a track with a fairly high energy level
- (Faceted) Tag: `(facet id, label, score)`, e.g. `facet` = "genre", `label` = "Hip Hop/Rap", `score` = 0.3 for tagging a track that resembles the genre "Hip Hop/Rap" to a fairly low amount and might be a candidate during a genre change. The "genre" tag with the highest score will represent the *main* genre and the score defines an ordering between multiple tags of the same *facet id*.

The tuple (facet id, label) defines a uniqueness constraint among all facets, i.e. within an instance of `Facets` which is basically two nested maps. The 1st key is *facet id* (nullable) and the 2nd key is *label* (nullable).

A new `Facets` property is supposed to be added to `Track`, but usage is not limited to that.

uklotzde committed:
Add faceted custom tags foundation with JSON serialization

Be-ing commented:
What remains to be done for this? Why is it still a draft pull request?

uklotzde commented:
This

> What remains to be done for this? Why is it still a draft pull request?

### Code structure
How should we organize the new code that will include database access, file import/export and UI components? I am proposing to put everything in a separate directory, i.e. slicing *vertically* like in Self-contained Systems (SCS) instead of *horizontally* by layers. This helps to isolate the contextual aspects and requires clear boundaries and interfaces. Not like a *bounded context* in DDD, more like a *toolbox* whose items can be instantiated and used wherever needed.

### Naming
Currently, the common directory is named `tagging`, but I am still not really happy with this choice. The directory name doesn't introduce a namespace, everything is placed in `mixxx`.

Be-ing commented:
> Currently, the common directory is named tagging, but I am still not really happy with this choice. The directory name doesn't introduce a namespace, everything is placed in mixxx.

Maybe move it to src/track/tagging? At a quick glance it doesn't look like this code has any inherit dependency on Track, but IIUC it isn't useful without associating it with tracks.

Be-ing commented:
> How should we organize the new code that will include database access, file import/export and UI components? I am proposing to put everything in a separate directory, i.e. slicing vertically like in Self-contained Systems (SCS) instead of horizontally by layers. This helps to isolate the contextual aspects and requires clear boundaries and interfaces. Not like a bounded context in DDD, more like a toolbox whose items can be instantiated and used wherever needed.

I don't really care either way. :shrug:

uklotzde commented:
Now in `mixxx::library::tags`.

Note: The JSON (de-)serialization is part of the classes for convenience and consistency. Technically, it would be preferable to split it off. But where to place and how to name those utility functions?

Be-ing commented:
In #2618 I separated the classes for serializing and deserializing data to XML from the classes which utilize that data. The (de)serialization classes are strictly read only and their members can only be set by the constructors. I think the lack of this clear separation of responsibilities is a major reason why the current architecture of the effects system in master got so convoluted with duplicated states. I think it would likely be good to apply the same approach here.

uklotzde commented:
> In #2618 I separated the classes for serializing and deserializing data to XML from the classes which generate and utilize that data. The (de)serialization classes are strictly read only and their members can only be set by the constructors. I think the lack of this clear separation of responsibilities is a major reason why the current architecture of the effects system in master got so convoluted with duplicated states. I think it would likely be good to apply the same approach here.

I don't see any such issues here. The serialization does not compromise the data structures. It either creates fresh, new instances when deserializing or serializes immutable instance. It's only about organizing the code and naming the functions.

uklotzde commented:
...and please note that this is pure data. No QObjects with signal/slot connections.

uklotzde committed:
Extract utility functions for parsing JSON data

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into custom-tags-foundation

uklotzde committed:
CustomTags: Log and report parsing errors

uklotzde commented:
Keeping the JSON serialization in the classes avoids unnecessary complexity until the separation is actually needed -> YAGNI

Be-ing commented on src/library/tags/tag.h:
```diff
@@ -0,0 +1,310 @@
+
+namespace tags {
+
+/// A category for tags.
```
Please clarify that this is a Mixxx feature and not to be conflated with file metadata tags.

uklotzde replied:
I have added the explanation in the docs for `Tag`. Might be moved somewhere else if we find a better place.

Be-ing commented on src/library/tags/tag.h:
```diff
@@ -0,0 +1,310 @@
+/// A category for tags.
+///
+/// Constraints: Lowercase ASCII, no whitespace characters
+class TagFacet final {
```
Please split these classes into separate files. I got confused at because I expected this class to be `Tag` matching the name of the file.

Be-ing commented on src/library/tags/tag.h:
```diff
@@ -0,0 +1,310 @@
+
+    /// Converts the given string into lowercase and then
+    /// removes all whitespace and non-ASCII characters.
+    static value_t clampValue(
```
`clamp` is not a verb whose meaning is clear in the context of operating on a string. Some alternative suggestions:
  * sanitizeValue
  * removeInvalidChars
  * convertToValidChars

Be-ing commented on src/library/tags/tag.h:
```diff
@@ -0,0 +1,310 @@
+    static bool isValidValue(
+            const value_t& value);
+
+    static value_t clampValue(
```
same comment about naming of this function applies as with TagFacet

Be-ing commented on src/library/tags/tag.h:
```diff
@@ -0,0 +1,310 @@
+    value_t m_value;
+};
+
+// A plain tag with label and score.
```
```suggestion
/// A plain tag with label and score.
```

Be-ing reviewed (changes requested):
Some initial comments just from reading the headers. I have not reviewed the implementations.

uklotzde committed:
Custom tags: Refine and extend documentation

uklotzde committed:
Custom tags: Rename member function

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into custom-tags-foundation

uklotzde commented on src/library/tags/customtags.h:
```diff
@@ -0,0 +1,196 @@
+///
+/// Each custom tag is represented by a triple: a facet, a label, and a score:
+///
+///   - The *facet* is a lowercase, non-empty ASCII string without whitespace
```
Explained in the docs now. Technically Unicode would be possible, but it is restricted on purpose.

uklotzde committed:
Tag facet/label/score: clampValue() -> convertIntoValidValue()

uklotzde committed:
Tag facet/label: Trade constexpr for validation

Use debug assertions to validate the string values, even though that
precludes constexpr.

uklotzde committed:
Custom tags: Split source files

uklotzde committed:
TagFacet: Enable definition of static, non-validated constants

uklotzde committed:
Explain the purpose of facet identifiers

uklotzde committed:
Fix clazy warnings

uklotzde committed:
Restrict the alphabet of tag facets

uklotzde committed:
Update tag facet doc comment

uklotzde committed:
Renaming: TagFacet -> TagFacetId

uklotzde committed:
Tags: Add constants for predefined facet identifiers

uklotzde committed:
Fix naming of constant

uklotzde committed:
Add more predefined tag facets and labels

uklotzde committed:
Rename CustomTags as Facets

uklotzde committed:
Remove redundant "Tag" prefix from type names

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into custom-tags-foundation

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into custom-tags-foundation

uklotzde committed:
Add utility functions and accompanying types for tag labels

Holzhaus commented on src/library/tags/facetid.cpp:
```diff
@@ -0,0 +1,176 @@
+namespace {
+
+const QRegularExpression kValidFacetStringNotEmpty(
+        QStringLiteral("^[\\+\\-\\./0-9@a-z\\[\\]_]+"));
```
Should we enforce starting with lowercase alphabetic letter? I think facet IDs like `_` would be weird. Just my 2 cents. 

```suggestion
        QStringLiteral("^[a-z][\\+\\-\\./0-9@a-z\\[\\]_]*"));
```

uklotzde committed:
Restrict facet identifiers to start with a lowercase alphabetic letter

uklotzde commented:
Closing due to lack of motivation. It is probably better to leave the Mixxx library as is. I am not keen on a review that lasts months or years with exhausting discussions. Please do not reopen!
